src/ComputeSharp.D2D1/Shaders/Interop/Effects/TransformMappers/D2D1TransformMapperImpl.cs

```diff
@@ -0,0 +1,286 @@
+        lpVtbl[4] = (void*)Marshal.GetFunctionPointerForDelegate(MapOutputRectToInputRectsWrapper);
+        lpVtbl[5] = (void*)Marshal.GetFunctionPointerForDelegate(MapInvalidRectWrapper);
+#endif
+
+        return lpVtbl;
+    }
+
+    /// <summary>
+    /// The vtable pointer for the current instance.
+    /// </summary>
+    private void** lpVtbl;
+
+    /// <summary>
+    /// The current reference count for the object (from <see cref="IUnknown"/>).
+    /// </summary>
+    private volatile int referenceCount;
+
+    /// <summary>
+    /// A <see cref="GCHandle"/> to the wrapped <see cref="ID2D1TransformMapperInterop"/> instance.
+    /// </summary>
+    private GCHandle transformMapperHandle;
+
+    /// <summary>
+    /// The <see cref="System.Threading.SpinLock"/> instance to synchronize accesses to this instance.
+    /// </summary>
+    private SpinLock spinLock;
+
+    /// <summary>
+    /// The factory method for <see cref="D2D1TransformMapperImpl"/> instances.
+    /// </summary>
+    /// <param name="transformMapper">The input <see cref="ID2D1TransformMapperInterop"/> instance to wrap.</param>
+    /// <param name="d2D1TransformMapperProxy">The resulting <see cref="D2D1TransformMapperImpl"/> instance.</param>
+    public static void Factory(
+        ID2D1TransformMapperInterop transformMapper,
+        D2D1TransformMapperImpl** d2D1TransformMapperProxy)
+    {
+        D2D1TransformMapperImpl* @this = (D2D1TransformMapperImpl*)NativeMemory.Alloc((nuint)sizeof(D2D1TransformMapperImpl));
+
+        @this->lpVtbl = Vtbl;
+        @this->referenceCount = 1;
+        @this->transformMapperHandle = GCHandle.Alloc(transformMapper);
+        @this->spinLock = new SpinLock(enableThreadOwnerTracking: false);
+
+        *d2D1TransformMapperProxy = @this;
+    }
+
+    /// <summary>
+    /// Gets a reference to the <see cref="System.Threading.SpinLock"/> instance used to synchronize reference counts.
+    /// </summary>
+    [UnscopedRef]
+    public ref SpinLock SpinLock => ref this.spinLock;
+
+    /// <summary>
+    /// Ensures the target of the current CCW is tracked and is keeping it alive.
+    /// </summary>
+    /// <param name="target">The associated <see cref="ID2D1TransformMapperInterop"/> object to track.</param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void EnsureTargetIsTracked(ID2D1TransformMapperInterop target)
+    {
+        this.transformMapperHandle.Target = target;
+    }
+
+    /// <summary>
+    /// Copies the current object onto a target pointer.
+    /// </summary>
+    /// <param name="transformMapper">The target <see cref="ID2D1TransformMapper"/> pointer.</param>
+    /// <returns>The <see cref="HRESULT"/> for the operation.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void CopyToWithNoLock(ID2D1TransformMapper** transformMapper)
+    {
+        this.referenceCount++;
```

Comment:
`referenceCount` may be `volatile`, but that doesn't make incrementing it with `++` _atomic_. For that you need `Interlocked.Increment`

Reply:
Not protecting the ref count with interlocked is how we get problems like D2D effect author objects causing crashes on the finalizer thread. Very bad. Definitely use interlocked for the reference count.

Reply:
This is fine, as the caller will already have taken a lock, so we don't need an interlocked increment. I've added more remarks in the XML docs for the two "NoLock" methods to further clarify this 👍